### PR TITLE
shift fix type coercion for integer64

### LINKED ISSUE #4865: shift() with data type integer64
Hello data.table team,

Here's an issue with the shift operator when used with data type integer64.

With type integer there are no issues:
```
d1 <- data.table(a=1:4)
d1[,b:=shift(a,fill=0L)]
d1[]                                        ## works as expected
   a b
1: 1 0
2: 2 1
3: 3 2
4: 4 3

```
Now with type integer64:
```
library(bit64)
d2 <- data.table(a=as.integer64(1:4))
d2[,b:=shift(a,fill=as.integer64(0))]        ## as.integer64("0") or as.integer64(0L) does not work either
Error in shift(a, fill = as.integer64(0)) : 
  INTEGER() can only be applied to a 'integer', not a 'double'

```
Strictly speaking, the data type of whatever we're filling with should be of the type of the rest of the vector.  But this works:
```
d2 <- data.table(a=as.integer64(1:4))
d2[,b:=shift(a,fill=0L)]
d2[]
   a b
1: 1 0
2: 2 1
3: 3 2
4: 4 3

> sapply(d2,class)
          a           b 
"integer64" "integer64" 
```
Maybe it should throw an error instead?

```
> sessionInfo()
R version 4.0.2 (2020-06-22)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18363)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] bit64_4.0.5       bit_4.0.4         data.table_1.13.0

loaded via a namespace (and not attached):
[1] compiler_4.0.2

```
I have the same issue on a linux box with older versions of R and data.table.

### PR DESCRIPTION
**Pull request by ben-schwen**
Closes #4865 

Also changes type coercion for the `integer` case to switch neatless between `integer`, `double` and `integer64`.

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -376,6 +376,8 @@
 
 44. In v1.13.2 a version of an old bug was reintroduced where during a grouping operation list columns could retain a pointer to the last group. This affected only attributes of list elements and only if those were updated during the grouping operation, [#4963](https://github.com/Rdatatable/data.table/issues/4963). Thanks to @fujiaxiang for reporting and @avimallu and Václav Tlapák for investigating and the PR.
 
+45. `shift(x, fill)` could error if `x` had type `bit64::integer64` and a `fill` argument was also provided, [#4865](https://github.com/Rdatatable/data.table/issues/4865). Thanks to @peterlittlejohn for reporting and Benjamin Schwendinger for the PR.
```
Comment (MichaelChirico) [inline NEWS.md:379]: Does  `could` really mean `would` here? If so, maybe reword:

> `shift(x, fill)` no longer fails when `x` is `integer64` and `fill` is provided

`could` makes it seem like sometimes fails, sometimes doesn't, in which case we'd want to explicate better what those failure cases are (and are not)

**Review by MichaelChirico — APPROVED** [review]

**Comment by MichaelChirico** [comment]
NB: coverage decrease is spurious

File: NEWS.md (hunk 1 above)
Comment (ben-schwen) [reply NEWS.md]: In this case its really a `could`, since it worked out `x` being `bit64::integer64` and `fill` being `integer32`. I tweaked the message a bit, but I'm not sure if `integer32` is an appropriate wording for NEWS.

**Comment by mattdowle** [comment]
I tackled the news item by focusing it on the most easily conveyed example: `nafill(xInt64, fill=0)`. Also added the error message so that folk searching for the error message might find it, as per contributing guide point 2 which has gotten quite dense now.
I didn't see why there were two blocks for the tests:  first one being one loop, the 2nd being a nested loop.  I started playing with that and testing in v1.14.2 to see which of the new tests worked before.  I liked the nested loop so tried to make that the only loop to see what would happen.  The revised new test passes in v1.14.2 but now fails in this PR.  I committed anyway, and I'm thinking `coerceAs` could be enhanced now to make these pass again, since it seems to be singleton 0 and NA which could be coerced without warning from complex.  Maybe there'll still be some warnings which should be tested by the tests. I'll keep going on this, hence WIP.
